diff --git a/master/buildbot/buildslave/ec2.py b/master/buildbot/buildslave/ec2.py
index 13e5a08a6..aa7dd4633 100644
--- a/master/buildbot/buildslave/ec2.py
+++ b/master/buildbot/buildslave/ec2.py
@@ -476,6 +476,7 @@ class EC2LatentBuildSlave(AbstractLatentBuildSlave):
                 bid_price = self.max_spot_price
         log.msg('%s %s requesting spot instance with price %0.2f.' %
                 (self.__class__.__name__, self.slavename, bid_price))
+        image = self.get_image()
         reservations = self.conn.request_spot_instances(
             bid_price, self.ami, key_name=self.keypair_name,
             security_groups=self.classic_security_groups,
@@ -489,7 +490,7 @@ class EC2LatentBuildSlave(AbstractLatentBuildSlave):
         instance_id = request.instance_id
         reservations = self.conn.get_all_instances(instance_ids=[instance_id])
         self.instance = reservations[0].instances[0]
-        return self._wait_for_instance(self.get_image())
+        return self._wait_for_instance(image)

     def _wait_for_instance(self, image):
         log.msg('%s %s waiting for instance %s to start' %
